File: day1/1.py

def read_file():
    f = open('./in', 'r')
    content = []
    for x in f:
        content.append(x.strip('\n'))
    return content


def two_entries(nums, target):
    seen = {}
    for n in nums:
        if (target - n) in seen:
            return n * (target -  n)
        else:
            seen[n] = True
    return None


def three_entries(nums, target):
    for i in range(0, len(nums)):
        remaining = target - nums[i]
        matched = two_entries(nums[i+1: len(nums)], remaining)
        if matched != None:
            return matched * nums[i]
    return -1
                

nums = list(map(int, read_file()))

print(two_entries(nums, 2020))
print(three_entries(nums, 2020))

File: day1/1.js
const fs = require('fs');


function two_entries(nums, target){
  const seen = {}
  for(let i = 0; i<nums.length; i++) {
    const complement = target - nums[i]
    if (seen[complement]) {
      return nums[i] * complement
    }
    seen[nums[i]] = true
  }
  return null
}


function three_entries(nums, target){
  for(let i = 0; i<nums.length; i++){
    const matched = two_entries(nums.slice(i+1, nums.length), target - nums[i])
    if(matched) {
      return matched * nums[i]
    }
  }
  return null;
}


fs.readFile('./in', 'utf8', (err, data) => {
  performCalc(data.split('\n').map(Number))
});


function performCalc(nums) {
  console.log(two_entries(nums, 2020))
  console.log(three_entries(nums, 2020))
}
